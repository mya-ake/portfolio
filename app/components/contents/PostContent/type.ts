export namespace PostContentType {
  export type Props = {
    slug: string;
  };
}
