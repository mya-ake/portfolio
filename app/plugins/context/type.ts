import { IHttp } from '@/lib/http';

export type Context = {
  http: IHttp<undefined>;
};
