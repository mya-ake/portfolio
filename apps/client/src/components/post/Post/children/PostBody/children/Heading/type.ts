import type { ReactNode } from 'react';

export type HeadingProps = {
  children: ReactNode;
};
