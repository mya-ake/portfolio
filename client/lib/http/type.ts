import { HTTP_METHODS } from './constants';

export type HttpMethods = typeof HTTP_METHODS[number];

export interface IHttp<E> {
  request<D>(
    request: HttpRequest,
    option?: HttpRequestOption,
  ): Promise<HttpResponse<D, E>>;

  cancelAll(): void;
  cancel(url: string): void;
}

export type HttpRequest = {
  url: string;
  method: HttpMethods;
};

export type HttpRequestOption = {
  cancelable: boolean;
};

interface IHttpResponse<D> {
  ok: boolean;
  status: number | undefined;
  data: D;
  noResponse: boolean;
  canceled: boolean;
}

export interface HttpOkResponse<D> extends IHttpResponse<D> {
  ok: true;
  status: number;
  noResponse: false;
  canceled: false;
}

export interface HttpErrorResponse<D> extends IHttpResponse<D> {
  ok: false;
  status: number;
  noResponse: false;
  canceled: false;
}

export interface HttpNoResponse extends IHttpResponse<undefined> {
  ok: false;
  status: undefined;
  data: undefined;
  noResponse: true;
  canceled: false;
}

export interface HttpCancelResponse extends IHttpResponse<undefined> {
  ok: false;
  status: undefined;
  data: undefined;
  noResponse: true;
  canceled: true;
}

export type HttpResponse<D, E> =
  | HttpOkResponse<D>
  | HttpErrorResponse<E>
  | HttpNoResponse
  | HttpCancelResponse;
