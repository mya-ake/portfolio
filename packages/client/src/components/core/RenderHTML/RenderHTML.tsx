import { createElement, Fragment, VFC, ReactElement } from 'react';
import { v4 as uuid } from 'uuid';
import type { Node as HTMLNode, TagNode } from '@mya-ake-com/parser';

type Render = (
  nodes: HTMLNode[],
  replacer?: Replacer,
) => ReactElement<unknown, string>;

type Replacer = (args: {
  tagNode: TagNode;
  childNodes: HTMLNode[];
  attrs: Record<string, string>;
  render: Render;
}) => ReactElement<unknown, string> | null | undefined;

export type RenderHTMLProps = {
  htmlNodes: HTMLNode[];
  replacer?: Replacer;
};

const render: Render = (nodes, replacer?) => {
  const elements = nodes.map((node) => {
    switch (node.nodeType) {
      case 'tag': {
        const attrs: Record<string, string> = {
          ...node.attrs,
          key: uuid(),
        };
        if ('class' in attrs) {
          attrs.className = node.attrs.class;
          delete attrs.class;
        }

        const replacerResult = replacer?.({
          tagNode: node,
          childNodes: node.childNodes,
          attrs,
          render: render,
        });
        if (replacerResult) {
          return replacerResult;
        }

        switch (node.tagName) {
          case 'img':
            return createElement(node.tagName, attrs);
          case 'script':
          case 'style':
            return null;
          default:
            return createElement(node.tagName, attrs, render(node.childNodes));
        }
      }
      case 'text':
        return createElement(Fragment, { key: uuid() }, node.content);
      default:
        return null;
    }
  });
  return <>{elements}</>;
};

export const RenderHTML: VFC<RenderHTMLProps> = ({ htmlNodes, replacer }) => {
  return render(htmlNodes, replacer);
};
