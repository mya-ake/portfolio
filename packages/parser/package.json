{
  "name": "@mya-ake-com/parser",
  "version": "1.0.0",
  "description": "parsers",
  "author": "mya-ake <make.app.ts@gmail.com> (https://github.com/mya-ake)",
  "homepage": "https://github.com/mya-ake/portfolio#readme",
  "license": "MIT",
  "main": "dist/index.js",
  "private": true,
  "repository": {
    "type": "git",
    "url": "git+https://github.com/mya-ake/portfolio.git"
  },
  "scripts": {
    "prebuild": "rimraf ./dist",
    "build": "tsc",
    "lint": "eslint --config ./.eslintrc.js .",
    "lint:fix": "eslint --config ./.eslintrc.js --fix .",
    "format": "prettier --write .",
    "format:check": "prettier --check .",
    "test:unit": "jest"
  },
  "bugs": {
    "url": "https://github.com/mya-ake/portfolio/issues"
  },
  "dependencies": {
    "parse5": "^7.0.0"
  },
  "devDependencies": {
    "@types/parse5": "^6.0.0"
  }
}
