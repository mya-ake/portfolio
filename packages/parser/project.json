{
  "root": "packages/parser/",
  "sourceRoot": "packages/parser/src",
  "targets": {
    "format-check": {
      "executor": "@nrwl/workspace:run-script",
      "options": {
        "script": "format:check"
      }
    },
    "lint-fix": {
      "executor": "@nrwl/workspace:run-script",
      "options": {
        "script": "lint:fix"
      }
    }
  }
}
