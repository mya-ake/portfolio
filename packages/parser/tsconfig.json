{
  "compilerOptions": {
    "target": "ES2018",
    "module": "CommonJS",
    "lib": ["esnext", "DOM"],
    "types": [],
    "strict": true,
    "moduleResolution": "node",
    "allowSyntheticDefaultImports": true,
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "outDir": "dist",
    "resolveJsonModule": true,
    "declaration": true
  },
  "include": ["./src/**/*.ts"],
  "exclude": ["./src/**/*.spec.ts"]
}
