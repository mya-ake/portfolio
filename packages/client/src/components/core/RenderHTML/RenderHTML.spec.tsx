import { render } from '@testing-library/react';
import { parseHtml } from '@mya-ake-com/parser';
import { RenderHTML, RenderHTMLProps } from './RenderHTML';
import type { FC, ReactNode } from 'react';

type WithChidlrenProps = {
  children: ReactNode;
};

type Replacer = Required<RenderHTMLProps>['replacer'];

describe('tags', () => {
  it('div', () => {
    const { container } = render(
      <RenderHTML htmlNodes={parseHtml(`<div>content</div>`)} />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('p with class', () => {
    const { container } = render(
      <RenderHTML htmlNodes={parseHtml(`<p class="text">content</p>`)} />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('p with em', () => {
    const { container } = render(
      <RenderHTML htmlNodes={parseHtml(`<p><em>content</em></p>`)} />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('a', () => {
    const { container } = render(
      <RenderHTML
        htmlNodes={parseHtml(`<a href="/to-link" rel="noopener">content</a>`)}
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('img', () => {
    const { container } = render(
      <RenderHTML
        htmlNodes={parseHtml(
          `<img src="https://example.com/sample.png" alt="sample">`,
        )}
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('br', () => {
    const { container } = render(<RenderHTML htmlNodes={parseHtml(`<br/>`)} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('ul li', () => {
    const { container } = render(
      <RenderHTML
        htmlNodes={parseHtml(`<ul><li>item 1</li><li>item 2</ul>`)}
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('not render tags', () => {
  it('script', () => {
    const { container } = render(
      <RenderHTML
        htmlNodes={parseHtml(`<script src="https://evil.example.com/evil.js">`)}
      />,
    );
    expect(container.firstChild).toBeNull();
  });

  it('inline script', () => {
    const { container } = render(
      <RenderHTML
        htmlNodes={parseHtml(`<script>alert('Hijacked');</script`)}
      />,
    );
    expect(container.firstChild).toBeNull();
  });

  it('style', () => {
    const { container } = render(
      <RenderHTML
        htmlNodes={parseHtml(`<style src="https://evil.example.com/evil.css">`)}
      />,
    );
    expect(container.firstChild).toBeNull();
  });

  it('inline style', () => {
    const { container } = render(
      <RenderHTML
        htmlNodes={parseHtml(
          `<style>input[value="a"]{ background-image: url('https://evil.example.com/?t=a'); }</style>`,
        )}
      />,
    );
    expect(container.firstChild).toBeNull();
  });
});

describe('replacer', () => {
  const Paragraph: FC<WithChidlrenProps> = ({ children }) => (
    <p className="p-0">{children}</p>
  );
  const Emphasis: FC<WithChidlrenProps> = ({ children }) => (
    <em className="text-bold">{children}</em>
  );

  it('p tag to Paragraph', () => {
    const replacer: Replacer = ({ tagNode, childNodes, attrs, render }) => {
      switch (tagNode.tagName) {
        case 'p':
          return <Paragraph {...attrs}>{render(childNodes)}</Paragraph>;
        default:
          return null;
      }
    };

    const { container } = render(
      <RenderHTML
        htmlNodes={parseHtml(`<p class="text">content</p>`)}
        replacer={replacer}
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('p tag to Paragraph, em tag to Emphasis', () => {
    const replacer: Replacer = ({ tagNode, childNodes, attrs, render }) => {
      switch (tagNode.tagName) {
        case 'p':
          return <Paragraph {...attrs}>{render(childNodes)}</Paragraph>;
        case 'em':
          return <Emphasis {...attrs}>{render(childNodes)}</Emphasis>;
        default:
          return null;
      }
    };

    const { container } = render(
      <RenderHTML
        htmlNodes={parseHtml(`<p><em>content</em></p>`)}
        replacer={replacer}
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
