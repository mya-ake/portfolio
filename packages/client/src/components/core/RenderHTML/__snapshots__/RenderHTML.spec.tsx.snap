// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`replacer p tag to Paragraph 1`] = `
<p
  class="p-0"
>
  content
</p>
`;

exports[`tags a 1`] = `
<a
  href="/to-link"
  rel="noopener"
>
  content
</a>
`;

exports[`tags br 1`] = `<br />`;

exports[`tags div 1`] = `
<div>
  content
</div>
`;

exports[`tags img 1`] = `
<img
  alt="sample"
  src="https://example.com/sample.png"
/>
`;

exports[`tags p with class 1`] = `
<p
  class="text"
>
  content
</p>
`;

exports[`tags ul li 1`] = `
<ul>
  <li>
    item 1
  </li>
  <li>
    item 2
  </li>
</ul>
`;
